blocks/api/utils.js

```diff
@@ -40,3 +42,45 @@ export function isUnmodifiedDefaultBlock( block ) {
 		isEqual( newDefaultBlock.attributes[ key ], block.attributes[ key ] )
 	);
 }
+
+/**
+ * Array of icon colors containing a color to be used if the icon color
+ * was not explicitly set but the icon background color was.
+ *
+ * @type {Object}
+ */
+const ICON_COLORS = [ '#191e23', '#fefdfe', '#7e70af' ];
+
+/**
+ * Function that receives an icon as set by the blocks during the registration
+ * and returns a new icon object that is normalized so we can rely on just on possible icon structure
+ * in the codebase.
+ *
+ * @param {(Object|string|WPElement)} icon  Slug of the Dashicon to be shown
+ *                                          as the icon for the block in the
+ *                                          inserter, or element or an object describing the icon.
+ *
+ * @return {Object} Object describing the icon.
+ */
+export function normalizeIconObject( icon ) {
+	if ( ! icon ) {
+		icon = { src: 'block-default' };
+	}
+	if ( isString( icon ) || icon instanceof Component ) {
+		icon = { src: icon };
+	}
+
+	if ( icon.backgroundColor && ( ! icon.iconColor || ! icon.shadowColor ) ) {
```

Comment:
Should we allow shadow color to be customizable?

Otherwise, we could simplify this to "background" and "foreground". `icon.iconColor` sounds a bit redundant.

Reply:
I followed this idea 👍

Reply:
> Otherwise, we could simplify this to "background" and "foreground". `icon.iconColor` sounds a bit redundant.

In retrospect, we should have done the same for `shadowColor` as well. Now we have some inconsistency with "`Color`" as only a suffix for shadow, when it's of the same value type as `foreground` and `background` (a hex string).

If we care enough, we could soft-deprecate this via something like:

```js
const icon = { ... };

Object.defineProperty( icon, 'shadowColor', {
	get() {
		deprecated( ... );

		return icon.shadow;
	},
} );
```